File: response_handler.cpp
#include "response.h"
/*
* This file contains methods of the class Response, to read and parse the server's responses according to protocol.
*/

// function to read a number of type T from the socket (assumes little endian representation)
template <typename T> T Response::readnum(std::string& messsage, uint32_t &offset) {
	T value = 0;
	std::memcpy(&value, messsage.data()+offset, sizeof(T));
	value = boost::endian::little_to_native<T>(value);
	offset += sizeof(T);
	return value;
}

// constructor for Response class. reads the server's response from the sokcet.
Response::Response(std::shared_ptr<tcp::socket>& socket)
	: head(std::make_shared<Response_Header>(socket)) {
	std::string message(head->payload_size, '\0');
	socket->read_some(boost::asio::buffer(message));

	// initialize body based on the response code
	switch (head->code) {
	case FILE_RECEIVED:
		body = std::make_shared<Valid_CRC_Response_Body>(message);
		break;
	case PUBLIC_KEY_RECEIVED:
	case SUCCESSFULL_RECONNECTION:
		body = std::make_shared<Response_Body_With_Key>(message);
		break;
	default:
		body = std::make_shared<Response_Body>(message);
	}
}

// prints the server's response code.
void Response::print_response_code() {
	if (Response::CODES_MEANING.find(head->code) == Response::CODES_MEANING.end())
		throw std::runtime_error("Server responded with unknown code: " + std::to_string(head->code) + ".\n");

	std::cout << "----------------------------------------------------------------\n";
	std::cout << "Server responded with code " << head->code << ':' << " \"" << Response::CODES_MEANING.at(head->code) << "\"\n";
	std::cout << "----------------------------------------------------------------\n\n";
}

//constructor for response header - reads the response header from the socket
Response::Response_Header::Response_Header(std::shared_ptr<tcp::socket>& socket) {
	std::string serialized_data(HEADER_SIZE, '\0');
	socket->read_some(boost::asio::buffer(serialized_data));
	uint32_t offset = 0;
	version = readnum<uint8_t>(serialized_data, offset);
	code = readnum<uint16_t>(serialized_data, offset);
	payload_size = readnum<uint32_t>(serialized_data, offset);
}

//constructors for different response body types
Response::Response_Body::Response_Body(std::string &body) {
	std::memcpy(client_id, body.data(), CLIENT_ID_SIZE);
}

Response::Response_Body_With_Key::Response_Body_With_Key(std::string &body) : Response_Body(body) {
	encrypted_key.resize(body.size() - CLIENT_ID_SIZE);
	std::memcpy(encrypted_key.data(), body.data() + CLIENT_ID_SIZE, encrypted_key.size());
}

Response::Valid_CRC_Response_Body::Valid_CRC_Response_Body(std::string &body) : Response_Body(body) {
	uint32_t offset = CLIENT_ID_SIZE;
	content_size = readnum<uint32_t>(body, offset);
	file_name.resize(NAME_LEN);
	std::memcpy(file_name.data(), body.data() + offset, NAME_LEN);
	file_name = std::string(file_name);
	offset += NAME_LEN;
	cksum = Response::readnum<uint32_t>(body, offset);
}

// getters
unsigned int Response::get_code() {
	return head->code;
}

unsigned char* Response::get_client_id() {
	return body->client_id;
}

unsigned long Response::get_cksum() {
	return body->get_cksum();
}

unsigned long Response::Response_Body::get_cksum() {
	return 0; // virtual getter, meant to be overriden if the body is of type Valid_CRC_Response_Body - returns junk value.
}

unsigned long Response::Valid_CRC_Response_Body::get_cksum() {
	return cksum;
}

std::string Response::get_aes_key() {
	if (head->code == PUBLIC_KEY_RECEIVED || head->code == SUCCESSFULL_RECONNECTION)
		return body->get_aes_key();
	else
		return "";
}

std::string Response::Response_Body::get_aes_key() {
	return ""; // virtual getter, meant to be overriden if the body is of type Response_Body_With_Key - returns junk value.
}

std::string Response::Response_Body_With_Key::get_aes_key() {
	return encrypted_key;
}

File: response_handler.py
'''
This file contains the Response class, that has methods to represent, serialize and send a response.
'''

import struct
DEFAULT_VERSION = 3

class Response:
    # response codes
    SUCCESSFUL_REGISTRATION = 1600
    REGISTRATION_FAILED = 1601
    PUBLIC_KEY_RECEIVED = 1602
    FILE_RECEIVED = 1603
    MESSAGE_RECEIVED = 1604
    SUCCESSFUL_RECONNECTION = 1605
    RECONNECTION_FAILED = 1606
    GENERAL_ISSUE = 1607
    codes = {
        1600: "SUCCESSFUL_REGISTRATION",
        1601: "REGISTRATION_FAILED",
        1602: "PUBLIC_KEY_RECEIVED",
        1603: "FILE_RECEIVED",
        1604: "MESSAGE_RECEIVED",
        1605: "SUCCESSFUL_RECONNECTION",
        1606: "RECONNECTION_FAILED",
        1607: "GENERAL_ISSUE"
    }

    def __init__(self, version, code, body):
        self.version = version
        self.code = code
        self.payload = body.serialize()
        self.payload_size = len(self.payload)

    def serialize(self):
        format_string = '<BHI'
        return struct.pack(format_string, self.version, self.code, self.payload_size) + self.payload

    def send_response(self, socket):
        print(f"Responding with code {self.code}: {self.codes[self.code]}.")
        socket.send(self.serialize())

    # classes to represent the different types of response bodies
    class Response_Body:
        def __init__(self, client_id=b''):  # default value for responses that don't contains a client id
            self.client_id = client_id

        def serialize(self):
            return self.client_id

    class Send_Key_Response_Body(Response_Body):
        def __init__(self, client_id, encrypted_key):
            super().__init__(client_id)
            self.encrypted_key = encrypted_key

        def serialize(self):
            return super().serialize()+self.encrypted_key

    class Send_File_Response_Body(Response_Body):
        def __init__(self, client_id, content_size, file_name, cksum):
            super().__init__(client_id)
            self.content_size = content_size
            self.file_name = file_name
            self.cksum = cksum

        def serialize(self):
            padded_file_name = self.file_name.encode('utf-8')[:255].ljust(255, b'\x00')
            format_string = '<I255sI'
            return super().serialize() + struct.pack(format_string, self.content_size, padded_file_name, self.cksum)

    # method to send a general error response
    @staticmethod
    def send_general_error(socket):
        resp = Response(DEFAULT_VERSION, Response.GENERAL_ISSUE, Response.Response_Body())
        resp.send_response(socket)
